File: freeq-android/freeq/src/main/java/com/freeq/model/KeyLookupCache.kt
package com.freeq.model

import com.freeq.ffi.KeyLookupStore
import java.io.File

/**
 * Where the SDK's key lookup keeps what it has learned between launches:
 * the signing keys it found, each account's proven records with their
 * listing time, and the proof CIDs that checked.
 *
 * One file per device, shared by every session on it, guests included, and
 * kept through sign-out, as the web app keeps its snapshot.
 * Nothing here is secret — every record in it is public in its account's
 * repository, and every proof was checked before it was kept — so it sits in
 * `filesDir` beside the buffer cache rather than in `securePrefs`, and needs
 * no account to name it.
 */
class AndroidKeyLookupStore(private val dir: File) : KeyLookupStore {

    private val file: File get() = File(dir, FILE_NAME)

    override fun load(): String? =
        try {
            if (file.exists()) file.readText() else null
        } catch (e: Exception) {
            android.util.Log.w("freeq.keylookup", "reading the key lookup cache failed", e)
            null
        }

    /**
     * Write through a temporary file and rename, so a process killed
     * mid-write leaves the previous snapshot intact rather than a half file.
     */
    override fun save(snapshot: String) {
        try {
            if (!dir.exists() && !dir.mkdirs()) return
            val tmp = File(dir, TMP_NAME)
            tmp.writeText(snapshot)
            if (!tmp.renameTo(file)) tmp.delete()
        } catch (e: Exception) {
            android.util.Log.w("freeq.keylookup", "keeping the key lookup cache failed", e)
        }
    }

    companion object {
        private const val FILE_NAME = "key-lookup.json"
        private const val TMP_NAME = "$FILE_NAME.tmp"
    }
}

File: freeq-ios/freeq/Models/KeyLookupCache.swift
import Foundation
import os.log

/// Where the SDK's key lookup keeps what it has learned between launches:
/// the signing keys it found, each account's proven records with their
/// listing time, and the proof CIDs that checked.
///
/// One file per device, shared by every session on it, guests included, and
/// kept through sign-out, as the web app keeps its snapshot.
/// Nothing here is secret — every record in it is public in its account's
/// repository, and every proof was checked before it was kept — so it sits
/// in Application Support beside the buffer cache rather than the Keychain,
/// which the device key needs and this does not, and needs no account to
/// name it.
final class FileKeyLookupStore: KeyLookupStore {

    private static let log = Logger(subsystem: "at.freeq.ios", category: "keylookup")
    private static let fileName = "key-lookup.json"

    private func url() -> URL? {
        Self.directory()?.appendingPathComponent(Self.fileName)
    }

    func load() throws -> String? {
        guard let url = url(), FileManager.default.fileExists(atPath: url.path) else { return nil }
        do {
            return try String(contentsOf: url, encoding: .utf8)
        } catch {
            Self.log.warning("reading the key lookup cache failed: \(error.localizedDescription)")
            return nil
        }
    }

    /// Write through a temporary file and replace, so a process killed
    /// mid-write leaves the previous snapshot intact rather than a half file.
    func save(snapshot: String) throws {
        guard let url = url() else { return }
        let temp = url.appendingPathExtension("tmp")
        do {
            try snapshot.write(to: temp, atomically: false, encoding: .utf8)
            if FileManager.default.fileExists(atPath: url.path) {
                _ = try FileManager.default.replaceItemAt(url, withItemAt: temp)
            } else {
                try FileManager.default.moveItem(at: temp, to: url)
            }
        } catch {
            try? FileManager.default.removeItem(at: temp)
            Self.log.warning("keeping the key lookup cache failed: \(error.localizedDescription)")
        }
    }

    private static func directory() -> URL? {
        let fm = FileManager.default
        guard let appSupport = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
        else { return nil }
        let dir = appSupport.appendingPathComponent("freeq", isDirectory: true)
        if !fm.fileExists(atPath: dir.path) {
            do { try fm.createDirectory(at: dir, withIntermediateDirectories: true) }
            catch { return nil }
        }
        return dir
    }
}
